### PR TITLE
Porous burner example

### PR DESCRIPTION
**Pull request by g3bk47**
Hi all,
I really like the extensible reactor classes. However, I feel that they are more powerful than what the currently available examples show. Therefore, I would like to contribute another example which uses the extensible reactors to simulate reacting flows in porous media burners. The porous media burner in question has been experimentally investigated and the burner setup as well as the governing equations are described in a recent publication (https://doi.org/10.1016/j.combustflame.2023.112642).
Since the code in this example uses the real properties and submodels for the different burner sections, it is quite a bit longer than other examples available in Cantera. I already simplified the code a bit compared to the actual code available in the supplementary materials of the paper, but it is still quite involved. The example showcases:
 - how to add a temperature equation for a solid phase;
 - add submodels for heat transfer between gas and solid, radiation, solid heat conduction;
 - describe the burner with an arbitrary amount of reactors;
 - describe the burner with an arbitrary amount of sections filled with different porous materials;
 - create a reactor cascade without mass flow controllers or pressure valves by assuming steady-state conditions and constant pressure.

Let me know what you think of this example and if I should simplify/shorten it further.

T

**Checklist**

- [X] The pull request includes a clear description of this code change
- [X] Commit messages have short titles and reference relevant issues
- [X] Build passes (`scons build` & `scons test`) and unit tests address code coverage
- [X] Style & formatting of contributed code follows [contributing guidelines](https://github.com/Cantera/cantera/blob/main/CONTRIBUTING.md)
- [X] The pull request is ready for review

### CODE REVIEW

File: data/YZA.yaml [hunk 1]
```diff
@@ -0,0 +1,26 @@
+description: |-
+  Properties for yttria-stabilzed zirconia alumina (YZA) used in
```
Comment (speth) [inline data/YZA.yaml:2]:
```suggestion
  Properties for yttria-stabilized zirconia alumina (YZA) used in
```

Comment (speth) [inline data/YZA.yaml]:
```suggestion
  Properties for yttria-stabilized zirconia alumina (YZA) used in
```

File: data/YZA.yaml [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+  porous media burners. Data compiled by Thorsten zirwes and used in
```
Comment (speth) [inline data/YZA.yaml]:
```suggestion
  porous media burners. Data compiled by Thorsten Zirwes and used in
```

File: data/YZA.yaml [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+  https://doi.org/10.1016/j.combustflame.2023.112642
+
+phases:
+- name: yttria_stabilzed_zirconia
```
Comment (speth) [inline data/YZA.yaml]: Besides the spelling, we generally use dashes instead of underscores in the YAML format
```suggestion
- name: yttria-stabilized-zirconia
```

File: AUTHORS [hunk 4]
```diff
@@ -59,6 +59,7 @@ Raymond Speth (@speth), Massachusetts Institute of Technology
 Su Sun (@ssun30), Northeastern University
 Sergey Torokhov (@band-a-prend)
 Laurien Vandewalle (@lavdwall)
+Guillaume Vignat, Stanford University
```
Comment (speth) [inline AUTHORS]: If Guillaume has a GitHub account, can you (a) add their account name here like it is for others?

Also, please amend the commit message for the commit that adds the example to contain a line formatted like:
```
Co-authored-by: Guillaume Vignat <username@stanford.edu>

File: samples/python/reactors/PorousMediaBurner.py [hunk 5]
```diff
@@ -0,0 +1,395 @@
...
+This heat is then transported by conduction in the solid and radiation upstream
+to heat up the reactants and stabilize the reaction zone.
+Given a reaction mechanism, equations for species mass
+fractions, gas phase temperature and solid temperature are solved.
+The reactor cascade consists of a variable number of cylindrical reactors.
+The Rosseland model is used for radiative heat transfer by the solid.
+A Nusselt number model is used to compute the heat transfer coefficient.
+This example assumes finding steady-state solutions with constant pressure
+so that mass controllers and pressure valves are not required. The code is written in
+a general way to support an arbitrary number of reactors and an arbirary number of
```
Comment (speth) [inline samples/python/reactors/PorousMediaBurner.py]:
```suggestion
a general way to support an arbitrary number of reactors and an arbitrary number of
```

File: samples/python/reactors/PorousMediaBurner.py [hunk 6]
```diff
@@ -0,0 +1,395 @@
...
+Tamb = 300.            # ambient temperature (used for radiation modeling) (K)
+TsInit = 300.          # initial temperature of the solid (K)
+constantP = ct.one_atm # gas pressure inside the porous medium (Pa)
+mech = "gri30.yaml"    # reaction mechanism file
+phi = 1.2              # equivalence ratio (phi) of the fuel/oxidizer mixture
+mdot = 0.3             # mass flux through the cascade (assumed constant) (kg/m^2/s)
+                       # with respect to the cross-sectional area of the burner
+
+# initial state of the gas phase containing fuel and oxidizer at the inlet
+def getInitalGas(gas):
```
Comment (speth) [inline samples/python/reactors/PorousMediaBurner.py]:
```suggestion
def getInitialGas(gas):
```
And of course all places where this is called.

File: samples/python/reactors/PorousMediaBurner.py [hunk 7]
```diff
@@ -0,0 +1,395 @@
...
+# Class that represents a reactor of the reactor cascade
+class ReactorProperties:
+    def __init__(self, diameter, length, midpoint, chemistry, TsInit, solid):
+
+        self.diameter = diameter   # diameter of the reactor (m)
+        self.length = length       # length of the section (m)
+        self.midpoint = midpoint   # coordinate of the reactor center (m)
+        self.chemistry = chemistry # chemistry on or off in the reactor
+        self.TsInit = TsInit       # initial temperature of the solid (K)
+        self.solid = solid         # poperties of the solid porous medium
```
Comment (speth) [inline samples/python/reactors/PorousMediaBurner.py]:
```suggestion
        self.solid = solid         # properties of the solid porous medium
```

File: samples/python/reactors/PorousMediaBurner.py [hunk 8]
```diff
@@ -0,0 +1,395 @@
...
+        #                             species mass fractions                          #
+        # ============================================================================#
+
+        wdot = self.kinetics.net_production_rates # chemical source terms
+        if not self.chemistry:
+            wdot *= 0.0
+
+        # right hand side of the mass fraction equations
+        for k in range(self.thermo.n_species):
+            index = self.component_index(self.thermo.species_name(k))
```
Comment (speth) [inline samples/python/reactors/PorousMediaBurner.py]: Calling `component_index` and `species_name` so many times in this inner loop is inefficient. You can get a noticeable speedup by just calculating the species offset once in `after_initialize` as
```py
species_offset = self.component_index(self.thermo.species_name(0))
```
and calculating `index` as `k + species_offset` here.

File: samples/python/reactors/PorousMediaBurner.py [hunk 9]
```diff
@@ -0,0 +1,395 @@
...
+            # chemical source term
+            RHS[index] += wdot[k] * self.thermo.molecular_weights[k] / density
```
Comment (speth) [inline samples/python/reactors/PorousMediaBurner.py]: Accessing each element of the `molecular_weights` array like this causes it to be retrieved from C++ in every single loop iteration. Storing the `molecular_weights` array as a member of the `PMReactor` object during initialization will help quite a bit.

File: samples/python/reactors/PorousMediaBurner.py [hunk 10]
```diff
@@ -0,0 +1,395 @@
...
+            # convective contribution
+            RHS[index] += self.A * mdot / \
+                (self.V * porosity * density) * (Y_in[k] - self.thermo.Y[k])
```
Comment (speth) [inline samples/python/reactors/PorousMediaBurner.py]: Same thing applies here for `self.thermo.Y`, though that of course needs to be evaluated within `replace_eval`.

File: samples/python/reactors/PorousMediaBurner.py [hunk 11]
```diff
@@ -0,0 +1,395 @@
...
+The reactor cascade consists of a variable number of cylindrical reactors.
+The Rosseland model is used for radiative heat transfer by the solid.
+A Nusselt number model is used to compute the heat transfer coefficient.
+This example assumes finding steady-state solutions with constant pressure
+so that mass controllers and pressure valves are not required. The code is written in
+a general way to support an arbitrary number of reactors and an arbirary number of
+burner sections with different physical properties.
+
+The implemented equations make use of Cantera's extensible reactor models.
+Therefore, Cantera version 2.6.0a4 or higher is required.
```
Comment (speth) [inline samples/python/reactors/PorousMediaBurner.py]:
```suggestion
Therefore, Cantera version 2.6.0 or higher is required.
```

File: samples/python/reactors/PorousMediaBurner.py [hunk 12]
```diff
@@ -0,0 +1,395 @@
...
+
+    Experimental and numerical investigation of flame stabilization and
+    pollutant formation in matrix stabilized ammonia-hydrogen combustion,
+    G. Vignat, T. Zirwes, E.R. Toro, K. Younes, E. Boigne, P. Muhunthan,
+    L. Simitz, D. Trimis, M. Ihme
+    Combustion and Flame (https://doi.org/10.1016/j.combustflame.2023.112642)
+
+Requires: cantera >= 2.6.0a4
+Keywords: porous media burner, heat transfer between gas and solid,
+          radiation modeling, reactor cascade
```
Comment (speth) [inline samples/python/reactors/PorousMediaBurner.py]: The required version should just be "2.6.0". You should also list `matplotlib` as a dependency (we usually put ">= 2.0" as the version requirement).

For the keywords, please see the list in `doc/example-keywords.txt` and add any that seem relevant. "user-defined model" is the one that covers the use of `ExtensibleReactor`. If you want to add any beyond the existing ones, they should be ones where there's some likelihood of (eventually) having more examples that use the same keywords. I think I'd suggest:
```
Keywords: user-defined model, reactor network, combustion, porous media, heat transfer, radiative heat transfer
```

**Review by speth — CHANGES_REQUESTED** [review]
Thanks for providing this example, @g3bk47. I definitely agree that capabilities of the `ExtensibleReactor` classes are not fully explored in the existing examples. Your use case here isn't something I anticipated, but I'm glad to see that you were able to use it effectively.

While this is somewhat longer than other examples, I think that's completely fine, and this is a really nice example of building a "device scale" model on top of Cantera without being too complex.

File: samples/python/reactors/PorousMediaBurner.py [hunk 13]
```diff
@@ -0,0 +1,395 @@
+"""
+Short description: reactor cascade model for reactive flows in inert porous media
+based on extensible reactors. Showcases adding a temperature equation for a
+solid-phase and custom heat transfer/radiation models.
+
+Code by Thorsten Zirwes and Guillaume Vignat
+Stanford University & Karlsruhe Institute of Technology (KIT)
+2023
+
+This code implements a reactor cascade model for the simulation of reactive
```
Comment (ischoegl) [inline samples/python/reactors/PorousMediaBurner.py:10]: Please add a disclaimer stating limitations of the model, as is uses drastic simplifications in order to convert it from a 1D problem to a 0D cascade.

File: samples/python/reactors/PorousMediaBurner.py [hunk 14]
```diff
@@ -0,0 +1,395 @@
...
+        self.pore_diameter = pore_diameter         # characteristic pore diameter in m
+        self.heat_conductivity = heat_conductivity # solid heat conductivity W/(m/K)
+        self.specific_area = specific_area         # specific area (1/m)
+        self.emissivity = emissivity               # emissivity of the solid
+        self.solid_phase = solid_phase             # solution object for computing
+                                                   # solid heat capacities and densities
+
+# Thermal conductivity of the porous medium as function of solid temperature (W/m/K)
+def effectiveConductivitySiC(Ts):  # for silicon carbide
+    return (1 - 0.84) * 1857.0 * Ts**(-0.5332)
```
Comment (ischoegl) [inline samples/python/reactors/PorousMediaBurner.py:134]: please add comments explaining the choice of parameters.

File: samples/python/reactors/PorousMediaBurner.py [hunk 15]
```diff
@@ -0,0 +1,395 @@
...
+    midpoint = sum_lengths + 0.5 * rL  # midpoint of the current reactor
+    sum_lengths += rL
+
+    if midpoint < total_length * 0.5:
+        # for the left half of the burner, fill all reactors with cold inert 40 PPI YZA
+        props = ReactorProperties(
+            diameter=2 * inch,
+            length=rL,
+            midpoint=midpoint,
+            chemistry=False,
```
Comment (ischoegl) [inline samples/python/reactors/PorousMediaBurner.py:356]: The description of the example should mention that the upstream section does not consider chemistry.

**Comment by g3bk47** [comment]
Hi Ray, thanks for the review! I think I incorporated all of your comments with the latest commit.

**Review by ischoegl — CHANGES_REQUESTED** [review]
Hi @g3bk47 ... thanks for contributing a nice example based on `ExtensibleReactor`.

I believe that details of the implementation are in good hands (@speth already commented). 

From my perspective, I wanted to make sure that the example description includes a disclaimer clearly stating limitations of the model. Porous burners are intrinsically 1D; while it is certainly legitimate to simplify for some applications (e.g. coarse approximations of temperatures or predictions of emissions), a reactor cascade has inherent limitations that will not capture the full range of behaviors. At the same time the example will capture excess enthalpy creating superadiabatic conditions that promote combustion beyond conventional flammability limits and/or at rates that are significantly higher than adiabatic flame speeds. 

The most interesting aspect of two-section burners is that they possess the ability to stabilize flames for a broad range of inlet mass fluxes due to internal heat recirculation across the interface: stabilization is accomplished by mismatched matrix properties, where the downstream section by itself promotes flash-back while the upstream section by itself promotes blow-off. The stabilization mechanism is based on solid conduction / radiation, where implementation details are extremely relevant if one wants to predict flashback or blow-off behavior. It is unrealistic to capture these physics with a coarsely refined 0D approach (or disabling reactions).

In a nutshell, the model is certainly ok for illustration purposes, but will not necessarily have predictive qualities when it comes to conditions that can be realized in experiments. Thus, it is prudent to clearly point out these limitations to avoid misleading conclusions by users

**Comment by g3bk47** [comment]
Hi @ischoegl,
Thanks for your comments! I added clarifications about the limitations of this simple model and included references for the heat conductivity of the solids.

**Review by ischoegl — DISMISSED** [review]
@g3bk47 ... thanks for the prompt updates!

**Review by speth — APPROVED** [review]
Thanks for the updates, @g3bk47. I think this looks good. I just pushed an update to squash a couple of the typo-fix commits together. We can merge as soon as the CI finishes.

**Comment by HadiEbrahimiHaratameh** [comment]
> Hi @g3bk47 ... thanks for contributing a nice example based on `ExtensibleReactor`.
> 
> I believe that details of the implementation are in good hands (@speth already commented).
> 
> From my perspective, I wanted to make sure that the example description includes a disclaimer clearly stating limitations of the model. Porous burners are intrinsically 1D; while it is certainly legitimate to simplify for some applications (e.g. coarse approximations of temperatures or predictions of emissions), a reactor cascade has inherent limitations that will not capture the full range of behaviors. At the same time the example will capture excess enthalpy creating superadiabatic conditions that promote combustion beyond conventional flammability limits and/or at rates that are significantly higher than adiabatic flame speeds.
> 
> The most interesting aspect of two-section burners is that they possess the ability to stabilize flames for a broad range of inlet mass fluxes due to internal heat recirculation across the interface: stabilization is accomplished by mismatched matrix properties, where the downstream section by itself promotes flash-back while the upstream section by itself promotes blow-off. The stabilization mechanism is based on solid conduction / radiation, where implementation details are extremely relevant if one wants to predict flashback or blow-off behavior. It is unrealistic to capture these physics with a coarsely refined 0D approach (or disabling reactions).
> 
> In a nutshell, the model is certainly ok for illustration purposes, but will not necessarily have predictive qualities when it comes to conditions that can be realized in experiments. Thus, it is prudent to clearly point out these limitations to avoid misleading conclusions by users

Hi,
I Thank @g3bk47 and @speth for the code and comments on it.
I think we should apply the Porous including adding the energy equation for the solid in 1D rather than 0D.
If the class ExtensibleReactor could be extended for 1D reactor model  (e.g. PFR), the network of the reactors such a way that a network of the PFR reactors are used, the results may be more realistic,
Thank you
Hadi
